PublicGitArchive/pga-create/cmd/pga-create/repack.go

```diff
@@ -21,76 +26,179 @@ type repackCommand struct {
 }
 
 func (c *repackCommand) Execute(args []string) error {
-	repack(c.Stdin, c.URL, c.Output)
-
+	processDump(c.Stdin, c.URL, c.Output, repack)
 	return nil
 }
 
-func repack(stdin bool, url, output string) {
+func processDump(stdin bool, url, output string, process func(*tar.Reader, io.Writer) int64) {
 	startTime := time.Now()
 	spin := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
 	spin.Start()
 	defer spin.Stop()
 
-	inputFile := dumpReader(stdin, url, spin)
-	var totalRead int64
-	inputFile = trackingReader{RealReader: inputFile, Callback: func(n int) {
-		totalRead += int64(n)
-		if totalRead%3 == 0 {
-			// supposing that mod 3 is random, this is a quick and dirty way to (/ 3) updates.
-			spin.Suffix = fmt.Sprintf(" %s", humanize.Bytes(uint64(totalRead)))
-		}
-	}}
-	gzf, err := gzip.NewReader(inputFile)
+	input := dumpReader(stdin, url, spin)
+	tracker := &trackingReader{RealReader: input, spin: spin}
+	gzf, err := gzip.NewReader(tracker)
 	if err != nil {
 		fail("opening gzip stream", err)
 	}
 	defer gzf.Close()
 	tarf := tar.NewReader(gzf)
-	processed := int64(0)
 
 	outf, err := createOutput(output)
 	if err != nil {
 		fail("creating output file", err)
 	}
+
 	gzw := gzip.NewWriter(outf)
-	tarw := tar.NewWriter(gzw)
+	processed := process(tarf, gzw)
+	fmt.Printf("\nRead      %s\nProcessed %s\nElapsed   %s\n",
+		humanize.Bytes(uint64(tracker.TotalRead)), humanize.Bytes(uint64(processed)), time.Since(startTime))
+
+	if err := gzw.Close(); err != nil {
+		fail("closing output gz file", err)
+	}
+
+	if err := outf.Close(); err != nil {
+		fail("closing output file", err)
+	}
+}
+
+type trackingReader struct {
+	RealReader io.ReadCloser
+	TotalRead  int64
+	spin       *spinner.Spinner
+}
+
+func (r *trackingReader) Read(p []byte) (n int, err error) {
+	n, err = r.RealReader.Read(p)
+	r.TotalRead += int64(n)
+	if r.TotalRead%3 == 0 {
+		// supposing that mod 3 is random, this is a quick and dirty way to (/ 3) updates.
+		r.spin.Suffix = fmt.Sprintf(" %s", humanize.Bytes(uint64(r.TotalRead)))
+	}
 
-	numTasks := 3
-	status := 0
-	i := 0
-	for header, err := tarf.Next(); err != io.EOF; header, err = tarf.Next() {
+	return n, err
+}
+
+func (r *trackingReader) Close() error {
+	return r.RealReader.Close()
+}
+
+func dumpReader(stdin bool, url string, spin *spinner.Spinner) io.ReadCloser {
+	if stdin {
+		fi, err := os.Stdin.Stat()
+		if err != nil {
+			fail("checking stat on stdin", err)
+		}
+
+		if fi.Mode()&os.ModeNamedPipe != 0 {
+			return os.Stdin
+		}
+	}
+
+	if url == "" {
+		envURL := os.Getenv("GHTORRENT_MYSQL")
```

Comment:
If possible I'd have all parameters in the `Command` struct instead of reading environment in several places. This makes it easier to notice all the tunable values. You can use `env` to specify the name of the environment variable in the argument label:

```go
URL    string `short:"l" long:"url" description:"Link to GHTorrent MySQL dump in tar.gz format. If empty (default), it find the most recent dump at GHTORRENT_MYSQL ?= http://ghtorrent-downloads.ewi.tudelft.nl/mysql/." env:"GHTORRENT_MYSQL"`
```